presto-main/src/test/java/com/facebook/presto/operator/TestGroupByHash.java

```diff
@@ -284,4 +291,120 @@ else if (type == BIGINT) {
+    {
+        // Create a page with positionCount >> expected size of groupByHash
+        int length = 1_000_000;
+        Block valuesBlock;
+        if (type == VARCHAR) {
+            valuesBlock = createStringSequenceBlock(0, length);
+        }
+        else if (type == BIGINT) {
+            valuesBlock = createLongSequenceBlock(0, length);
+        }
+        else {
+            throw new IllegalArgumentException("unsupported data type");
+        }
+        Block hashBlock = getHashBlock(ImmutableList.of(type), valuesBlock);
+        Page page = new Page(valuesBlock, hashBlock);
+        AtomicInteger counter = new AtomicInteger();
+        int yields = 0;
+
+        // test addPage
+        GroupByHash groupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(type),
+                new int[] {0},
+                Optional.of(1),
+                4,
+                JOIN_COMPILER,
+                x -> counter.incrementAndGet() % 2 == 0);
+
+        Work addPageWork = groupByHash.addPage(page);
+        while (!addPageWork.process()) {
+            yields++;
+        }
+        assertTrue(addPageWork.process());
+        assertEquals(length, groupByHash.getGroupCount());
+        assertGreaterThanOrEqual(yields, (int) log(length), "at least log(length) many rehash");
+
+        // test getGroupIds
+        counter.set(0);
+        yields = 0;
+        groupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(type),
+                new int[] {0},
+                Optional.of(1),
+                4,
+                JOIN_COMPILER,
+                x -> counter.incrementAndGet() % 2 == 0);
+
+        Work<GroupByIdBlock> getGroupIdsWork = groupByHash.getGroupIds(page);
+        while (!getGroupIdsWork.process()) {
+            yields++;
+        }
+        assertTrue(getGroupIdsWork.process());
+        assertEquals(length, groupByHash.getGroupCount());
+        assertGreaterThanOrEqual(yields, (int) log(length), "at least log(length) many rehash");
+        assertEquals(length, getGroupIdsWork.getResult().getPositionCount());
+    }
+
+    @Test
+    public void testDictionaryYield()
+            throws Exception
+    {
+        // Create a page with positionCount >> expected size of groupByHash
+        int dictionaryLength = 1_000;
+        int length = 2_000;
+        int[] ids = IntStream.range(0, dictionaryLength).toArray();
+        DictionaryId dictionaryId = randomDictionaryId();
+        Block valuesBlock = new DictionaryBlock(dictionaryLength, createStringSequenceBlock(0, length), ids, dictionaryId);
+        Block hashBlock = new DictionaryBlock(dictionaryLength, getHashBlock(ImmutableList.of(VARCHAR), valuesBlock), ids, dictionaryId);
+        Page page = new Page(valuesBlock, hashBlock);
+        AtomicInteger counter = new AtomicInteger();
+        int yields = 0;
+
+        // test addPage
+        GroupByHash groupByHash = createGroupByHash(
```

Comment:
The tests are a bit repetitive, wonder if they are similar enough to extract some common pattern or building blocks.